implementations/rust/ockam/ockam_identity/src/channel/decryptor_worker.rs

```diff
@@ -220,51 +208,41 @@ impl DecryptorWorker {
     }
 
     async fn handle_key_exchange_msg(
-        &mut self,
-        ctx: &mut <Self as Worker>::Context,
-        msg: Routed<<Self as Worker>::Message>,
-    ) -> Result<()> {
-        let reply = msg.return_route();
-        self.remote_route = reply;
+        mut self,
+        ctx: &mut <DecryptorWorker as Worker>::Context,
+        msg: Routed<<DecryptorWorker as Worker>::Message>,
+    ) -> Result<State> {
+        self.remote_route = msg.return_route();
         let payload = Vec::<u8>::decode(&msg.into_transport_message().payload)?;
-
         self.handle_key_exchange(ctx, Some(&payload), false).await
     }
 
     async fn handle_key_exchange(
-        &mut self,
-        ctx: &mut <Self as Worker>::Context,
+        mut self,
+        ctx: &mut <DecryptorWorker as Worker>::Context,
         payload: Option<&[u8]>,
-        // If it's the first message
-        first_run: bool,
-    ) -> Result<()> {
-        let state;
-        if let Some(s) = self.state_key_exchange.as_mut() {
-            state = s;
-        } else {
-            return Err(IdentityError::InvalidSecureChannelInternalState.into());
-        }
-
-        if let Some(payload) = payload {
+        initialization_run: bool,
+    ) -> Result<State> {
+        if let Some(incoming_payload) = payload {
             // Received key exchange message from remote channel, need to forward it to the local key exchange
             debug!(
                 "SecureChannel received KeyExchangeRemote at {}",
                 &self.addresses.decryptor_remote
             );
-            let exchanger = &mut state.key_exchanger;
-            let _ = exchanger.handle_response(payload).await?;
+            let exchanger = &mut self.key_exchanger;
+            let _ = exchanger.handle_response(incoming_payload).await?;
         }
 
         // If we'll need to generate another request
         let mut request_was_sent = false;
 
         // Key exchange hasn't been completed -> generate and send next request
-        if !state.key_exchanger.is_complete().await? {
+        if !self.key_exchanger.is_complete().await? {
             request_was_sent = true;
-            let payload = state.key_exchanger.generate_request(&[]).await?;
+            let payload = self.key_exchanger.generate_request(&[]).await?;
 
             // We should send first_responder_address only with first message from the initiator
-            let custom_payload = if self.role.is_initiator() && first_run {
+            let custom_payload = if self.role.is_initiator() && initialization_run {
```

Comment:
I think that we should even be able to move the `initialization_run` piece of state into the `KeyExchangeState` and not pass it around.